src/core/include/mp-units/bits/scaling.h

```diff
@@ -0,0 +1,229 @@
+ * is chosen such that it is of comparable precision as the representation type,
+ *
+ * @note This is a low-level facility. Neither the `From`, nor the `To` types of the scaling
+ * operation are actually constrained to be floating-point or even "floating-point like" types.
+ * All it represents is the scaling operation as implemented by multiplication with a floating-point
+ * representation of the scaling factor. This is also used whenever simultaneously scaling and
+ * converting between integer and floating-point types.
+ *
+ * @tparam Rep Representation type
+ */
+template<typename Rep>
+struct floating_point_scaling_traits {
+  template<Magnitude M, typename From>
+  static constexpr Rep scale_from(M, const From& value)
+  {
+    return floating_point_scaling_impl<M{}>::template scale<Rep>(value);
+  };
+
+  template<Magnitude M>
+  static constexpr auto scale(M, const Rep& value)
+  {
+    // for standard floating-point types, the result respresentation is always the same
+    return floating_point_scaling_impl<M{}>::template scale<Rep>(value);
+  }
+};
+
+
+template<Magnitude auto M>
+struct fixed_point_scaling_impl {
+  static constexpr Magnitude auto num = _numerator(M);
+  static constexpr Magnitude auto den = _denominator(M);
+  template<std::integral T>
+  static constexpr auto ratio = [] {
+    using U = long double;
+    return detail::fixed_point<T>(_get_value<U>(M));
+  }();
+
+  template<typename To, typename From>
+    requires std::integral<value_type_t<To>> && std::integral<value_type_t<From>>
+  static constexpr To scale(const From& value)
+  {
+    using U = std::common_type_t<value_type_t<From>, value_type_t<To>>;
+    if constexpr (_is_integral(M)) {
+      return static_cast<To>(static_cast<value_type_t<From>>(value) * _get_value<U>(num));
+    } else if constexpr (_is_integral(_pow<-1>(M))) {
+      return static_cast<To>(static_cast<value_type_t<From>>(value) / _get_value<U>(den));
+    } else {
+      return static_cast<To>(ratio<U>.scale(static_cast<value_type_t<From>>(value)));
+    }
+  }
+};
+
+
+template<typename Rep>
+struct fixed_point_scaling_traits {
+  template<Magnitude M, typename From>
+  static constexpr Rep scale_from(M, const From& value)
+  {
+    return fixed_point_scaling_impl<M{}>::template scale<Rep>(value);
+  };
+
+  template<Magnitude M>
+  static constexpr auto scale(M, const Rep& value)
+  {
+    // for standard integer types, the result respresentation is always the same
+    return fixed_point_scaling_impl<M{}>::template scale<Rep>(value);
+  }
+};
+
+
+template<typename T, typename Other = T>
+inline constexpr auto select_scaling_traits = [] {
+  if constexpr (requires {
+                  // we only check if the traits class is complete, not it's members; we do not want to fall-back
+                  // depending on the argument types
+                  { sizeof(mp_units::scaling_traits<T>) } -> std::integral;
```

Comment:
I'm not a standard lawyer, but my tests on godbolt suggest that the latter just tests if the type is declared, not if it is defined/complete.

Reply:
Right. Still, I am not sure if checking for `size of` gives us much more here, as this may be defined as just `{}` with no interface.

Anyway, checking the return type of `sizeof` probably is not needed here. Users should not declare functions with names of language keywords.